### PR TITLE
PXB 8.0.29 merge

### PR DESCRIPTION
**Pull request by altmannmarcelo**
https://jira.percona.com/browse/PXB-2755

### CODE REVIEW

File: storage/innobase/xtrabackup/src/xtrabackup.cc [hunk 1]
```diff
@@ -7092,6 +7246,22 @@ static void xtrabackup_prepare_func(int argc, char **argv) {
 
     srv_shutdown_state = SRV_SHUTDOWN_NONE;
 
+    /* PXB-2792 - Ensure we don't have ib_logfile's */
```
Comment (rahulmalik87) [inline storage/innobase/xtrabackup/src/xtrabackup.cc]: It is better to create seprate commit  for this changes as we have jira ticket also

File: storage/innobase/xtrabackup/test/inc/common.sh [hunk 2]
```diff
@@ -2,7 +2,11 @@ set -eu
 
 function xtrabackup()
 {
+  if [ -z ${WITH_RR+x} ]; then
```
Comment (rahulmalik87) [inline storage/innobase/xtrabackup/test/inc/common.sh]: Is this new option we are introducing? we should mention about it in run.sh --help

File: storage/innobase/xtrabackup/test/inc/keyring_kms.sh [hunk 3]
```diff
@@ -51,16 +51,16 @@ function cleanup_keyring() {
 
 function ping_kms()
 {
-  if [ -z ${KMS_REGION+x} ]; then
+  if [ -z ${KMS_REGION} ]; then
```
Comment (rahulmalik87) [inline storage/innobase/xtrabackup/test/inc/keyring_kms.sh]: is it to report failure when KMS_REGION is set to null? Also better to add quotes.

File: storage/innobase/dict/dict0dd.cc [hunk 4]
```diff
@@ -6362,13 +6349,8 @@ const char *dd_process_dd_tables_rec_and_mtr_commit(
 
   /* Get the se_private_id field. */
   field = (const byte *)rec_get_nth_field(
-<<<<<<< HEAD
-      rec, offsets, dd_tables->field_number("se_private_id") + DD_FIELD_OFFSET,
-=======
       nullptr, rec, offsets,
-      dd_object_table.field_number("FIELD_SE_PRIVATE_ID") + DD_FIELD_OFFSET,
->>>>>>> mysql-8.0.29
```
Comment (satya-bodapati) [inline storage/innobase/dict/dict0dd.cc]: Do you know why we changed from "FIELD_SE_PRIVATE_ID" to se_private_id?

File: storage/innobase/include/fsp0fsp.ic [hunk 5]
```diff
@@ -366,11 +366,6 @@ inline bool fsp_flags_is_valid(uint32_t flags) {
     return (false);
   }
 
-<<<<<<< HEAD
-#if FSP_FLAGS_POS_UNUSED != 15
-#error You have added a new FSP_FLAG without adding a validation check.
-#endif
```
Comment (satya-bodapati) [inline storage/innobase/include/fsp0fsp.ic]: This seems like a valid check. Is it removed by upstream? Did upstream add a new FSP flag?

File: include/boost_1_77_0/patches/boost/container/throw_exception.hpp [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+#endif
+
+#if defined(BOOST_HAS_PRAGMA_ONCE)
+#pragma once
+#endif
+
+#include <boost/container/detail/config_begin.hpp>
+#include <boost/container/detail/workaround.hpp>
+#include <boost/core/ignore_unused.hpp>
+
```
Comment (satya-bodapati) [inline include/boost_1_77_0/patches/boost/container/throw_exception.hpp]: why is this required by us (PXB) and not upstream?

Comment (satya-bodapati) [reply include/boost_1_77_0/patches/boost/container/throw_exception.hpp]: this as in, the entire file

File: sql/join_optimizer/join_optimizer.cc [hunk 7]
```diff
@@ -3449,8 +3452,11 @@ void CostingReceiver::ProposeNestedLoopJoin(
       const CachedPropertiesForPredicate &properties =
           edge->expr->properties_for_equijoin_conditions[join_cond_idx];
 
+      MY_COMPILER_DIAGNOSTIC_PUSH()
+      MY_COMPILER_GCC_DIAGNOSTIC_IGNORE("-Wunused-variable")
       const auto [already_applied_as_sargable, subsumed] =
           AlreadyAppliedAsSargable(condition, left_path, right_path);
+      MY_COMPILER_DIAGNOSTIC_POP()
```
Comment (satya-bodapati) [inline sql/join_optimizer/join_optimizer.cc]: this is upstream code and it didn't catch this error? :thinking:

File: storage/innobase/dict/dict0dd.cc [hunk 8]
```diff
@@ -442,10 +444,10 @@ dict_table_t *dd_table_create_on_dd_obj(const dd::Table *dd_table,
                                         const dd::String_type *schema_name,
                                         bool is_implicit, space_id_t space_id) {
   using namespace dict_name;
-  mem_heap_t *heap = mem_heap_create(1000);
+  mem_heap_t *heap = mem_heap_create(1000, UT_LOCATION_HERE);
 
   char table_name[MAX_SPACE_NAME_LEN + 1];
-  char tmp_schema[MAX_DATABASE_NAME_LEN + 1];
+  char tmp_schema[+1];
```
Comment (satya-bodapati) [inline storage/innobase/dict/dict0dd.cc]: is this typo?

File: storage/innobase/include/fsp0fsp.ic [hunk 9]
```diff
@@ -321,6 +321,7 @@ inline bool fsp_flags_is_valid(uint32_t flags) {
   bool has_data_dir = FSP_FLAGS_HAS_DATA_DIR(flags);
   bool is_shared = FSP_FLAGS_GET_SHARED(flags);
   bool is_temp = FSP_FLAGS_GET_TEMPORARY(flags);
+  bool is_encryption = FSP_FLAGS_GET_ENCRYPTION(flags);
```
Comment (satya-bodapati) [inline storage/innobase/include/fsp0fsp.ic]: before this change, how did this function compile without the variable `is_encryption`?

File: storage/innobase/include/log0recv.h [hunk 10]
```diff
@@ -742,9 +742,8 @@ void recv_reset_buffer();
 bool recv_sys_resize_buf();
 
 /** Parse log records from a buffer and optionally store them to a
-hash table to wait merging to file pages.
-@param[in]  checkpoint_lsn  the LSN of the latest checkpoint */
-void recv_parse_log_recs(lsn_t checkpoint_lsn);
+hash table to wait merging to file pages. */
+void recv_parse_log_recs();
```
Comment (satya-bodapati) [inline storage/innobase/include/log0recv.h]: are we now accessing checkpoint_lsn from global recv_sys data structure?

File: storage/innobase/xtrabackup/src/backup_copy.cc [hunk 11]
```diff
@@ -477,7 +477,7 @@ bool backup_file_print(const char *filename, const char *message, int len) {
   MY_STAT stat;
 
   memset(&stat, 0, sizeof(stat));
-  stat.st_mtime = my_time(0);
+  stat.st_mtime = time(nullptr);
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/src/backup_copy.cc]: my_time() vs time(nullptr) ?

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: storage/innobase/dict/dict0dd.cc [hunk 12]
```diff
@@ -447,7 +447,7 @@ dict_table_t *dd_table_create_on_dd_obj(const dd::Table *dd_table,
   mem_heap_t *heap = mem_heap_create(1000, UT_LOCATION_HERE);
 
   char table_name[MAX_SPACE_NAME_LEN + 1];
-  char tmp_schema[+1];
+  char tmp_schema[MAX_DATABASE_NAME_LEN + 1];
```
Comment (satya-bodapati) [inline storage/innobase/dict/dict0dd.cc]: oh, you fixed here? I dropped a comment before. Plz ignore then ;)

File: storage/innobase/xtrabackup/test/inc/keyring_kms.sh (hunk 3 above)
Comment (altmannmarcelo) [reply storage/innobase/xtrabackup/test/inc/keyring_kms.sh]: Yes, this is to skip the test if we have not SET the variables.

File: storage/innobase/dict/dict0dd.cc (hunk 4 above)
Comment (altmannmarcelo) [reply storage/innobase/dict/dict0dd.cc]: @satya-bodapati We/upstream did not change.
PXB gather the information as `dd_tables->field_number("se_private_id")`
Upstream create a `dd_object_table` - https://github.com/mysql/mysql-server/blob/mysql-8.0.29/storage/innobase/dict/dict0dd.cc#L5332 and uses that object to fetch the field number via `dd_object_table.field_number("FIELD_SE_PRIVATE_ID")` . From what I can see, we first added the optimization of fetching the field number as name rather than fixed number. Later upstream did the same in a different way but we never conciliated to their way. 
However, here what changed was the first parameter `nullptr` we pass to `rec_get_nth_field` as this is new work added as part of INSTANT ADD/DROP column commit.

File: storage/innobase/include/fsp0fsp.ic (hunk 5 above)
Comment (altmannmarcelo) [reply storage/innobase/include/fsp0fsp.ic]: This comes from https://github.com/altmannmarcelo/percona-xtrabackup/commit/e001f2a533c4b8edd26e8e1723f00ed4a71d80c7 They removed `#defines` from code.
This particular part change to a `static_assert` - https://github.com/mysql/mysql-server/blob/mysql-8.0.29/storage/innobase/include/fsp0fsp.ic#L376

File: include/boost_1_77_0/patches/boost/container/throw_exception.hpp (hunk 6 above)
Comment (altmannmarcelo) [reply include/boost_1_77_0/patches/boost/container/throw_exception.hpp]: Seems like we do not need this after all. It might have been fixed by other later change. Will remove it

File: storage/innobase/include/fsp0fsp.ic (hunk 9 above)
Comment (altmannmarcelo) [reply storage/innobase/include/fsp0fsp.ic]: This has required as https://github.com/mysql/mysql-server/blob/mysql-8.0.29/storage/innobase/include/fsp0fsp.ic#L376 brought over:
```
  /* Only single-table and general tablespaces and not temp tablespaces
  use the encryption clause. */
  if (is_encryption && (is_temp)) {
    return (false);
  }
```
To be on the safe-side, since we do support encryption of temp table, I've kept that if and the is_encryption flag check under `ifndef XTRABACKUP`

File: storage/innobase/include/log0recv.h (hunk 10 above)
Comment (altmannmarcelo) [reply storage/innobase/include/log0recv.h]: This comes from https://github.com/mysql/mysql-server/commit/85f9abf99075052964d6434737a7645c311b2990#diff-26f5d4b9822d8a85edd9ea715cde19e90347775be561b8919601cfcb3df8cb49 . We were referencing checkpoint_lsn as a parameter but not using it.

From what I can see, we only used it on 5.7 and just to print a debug message - https://github.com/mysql/mysql-server/blob/5.7/storage/innobase/log/log0recv.cc#L3247

File: storage/innobase/xtrabackup/src/backup_copy.cc (hunk 11 above)
Comment (altmannmarcelo) [reply storage/innobase/xtrabackup/src/backup_copy.cc]: yes, this came from https://github.com/mysql/mysql-server/commit/7230b2e10e89e8df920f83fc11f12b97937b3f26
> Change all calls my_time(0) to ::time(nullptr).

File: sql/join_optimizer/join_optimizer.cc (hunk 7 above)
Comment (altmannmarcelo) [reply sql/join_optimizer/join_optimizer.cc]: yes, this error is thrown if compiling with GCC-7

File: storage/innobase/xtrabackup/test/inc/common.sh (hunk 2 above)
Comment (altmannmarcelo) [reply storage/innobase/xtrabackup/test/inc/common.sh]: Done

File: include/boost_1_77_0/patches/boost/container/throw_exception.hpp (hunk 6 above)
Comment (satya-bodapati) [reply include/boost_1_77_0/patches/boost/container/throw_exception.hpp]: Thanks

File: storage/innobase/include/fsp0fsp.ic (hunk 5 above)
Comment (satya-bodapati) [reply storage/innobase/include/fsp0fsp.ic]: ok, if there is static_assert, then we are all good

File: storage/innobase/dict/dict0dd.cc (hunk 4 above)
Comment (satya-bodapati) [reply storage/innobase/dict/dict0dd.cc]: Thanks for clarification.

**Comment by satya-bodapati** [comment]
https://pxb.cd.percona.com/job/percona-xtrabackup-8.0-test-param/198/CMAKE_BUILD_TYPE=Debug,DOCKER_OS=ubuntu%3Afocal,XTRABACKUP_TARGET=innodb80/
https://pxb.cd.percona.com/job/percona-xtrabackup-8.0-test-param/198/CMAKE_BUILD_TYPE=Debug,DOCKER_OS=centos%3A7,XTRABACKUP_TARGET=innodb80/
https://pxb.cd.percona.com/job/percona-xtrabackup-8.0-test-param/198/CMAKE_BUILD_TYPE=Debug,DOCKER_OS=asan,XTRABACKUP_TARGET=innodb80/

**Review by satya-bodapati — APPROVED** [review]
@altmannmarcelo 
LGTM! I request to you test import export of old version datadir with newer 8.0.29 binaries.
I assume there is a new CFG version and PXB 8.0.29 should create an appropriate CFG version.

**Review by rahulmalik87 — APPROVED** [review]
LGTM
